### PR TITLE
tpm: add openssl tpm engine for TPM1.2

### PR DESCRIPTION
**Pull request by limeng-linux**
openssl tpm engine is an interface between openssl application and TPM
hardware. Base on it, we can process security feature with TPM IC
instead of openssl software solution.

Signed-off-by: Meng Li <Meng.Li@windriver.com>

### CODE REVIEW

File: recipes-tpm/openssl-tpm-engine/files/0001-create-tpm-key-support-well-known-key-option.patch [hunk 1]
```diff
@@ -0,0 +1,101 @@
+commit 16dac0cb7b73b8a7088300e45b98ac20819b03ed
+Author: Junxian.Xiao <Junxian.Xiao@windriver.com>
+Date:   Wed Jun 19 18:57:13 2013 +0800
+
+support well-known password in openssl-tpm-engine.
+
+CQID: WIND00422121 WIND00422123
```
Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/files/0001-create-tpm-key-support-well-known-key-option.patch]: Remove the CQID line.

File: recipes-tpm/openssl-tpm-engine/files/0002-libtpm-support-env-TPM_SRK_PW.patch [hunk 2]
```diff
@@ -0,0 +1,82 @@
+commit 16dac0cb7b73b8a7088300e45b98ac20819b03ed
+Author: Junxian.Xiao <Junxian.Xiao@windriver.com>
+Date:   Wed Jun 19 18:57:13 2013 +0800
+
+support reading SRK password from env TPM_SRK_PW
+
+CQID: WIND00422121 WIND00422123
```
Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/files/0002-libtpm-support-env-TPM_SRK_PW.patch]: ditto

File: recipes-tpm/openssl-tpm-engine/files/README [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+     To reset SRK and Owner password to well-known key, run "tpm_changeownerauth -r -s -o". 
+
+(2). Generate tpm hardware key: (add -z for using well-known key)
+     $ create_tpm_key rootkey.pem [-z]
+
+(3). Wrap software key into TPM storage:
+     $ openssl genrsa -out softkey.pem 1024
+     $ create_tpm_key -w softkey.pem -s 1024 rootkey.pem [-z]
+
+
```
Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/files/README]: Remove this empty line.

File: recipes-tpm/openssl-tpm-engine/files/README [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+==================================
+2. How to use tpm engine lib 
+==================================
+The name of tpm engine library is libtpm.so.
+There is a enhencement to the original opensource code.
```
Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/files/README]: s/a/an/

File: recipes-tpm/openssl-tpm-engine/openssl-tpm-engine_0.4.2.bb [hunk 5]
```diff
@@ -0,0 +1,53 @@
+DESCRIPTION = " OpenSSL secure engine based on TPM hardware"
+HOMEPAGE = "http://www.openssl.org/"
+SECTION = "libs/network"
+PR = "r0"
+LICENSE = "openssl"
+
+DEPENDS = "openssl trousers"
+RDEPENDS_${PN} = "tpm-tools"
```
Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/openssl-tpm-engine_0.4.2.bb]: If libtpm.so has a runtime dependency of openssl, add it here.

Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/openssl-tpm-engine_0.4.2.bb]: Why tpm-tools is listed here.

**Comment by limeng-linux** [comment]
there are 2 binary output as below:
create_tpm_key
libtpm.so.0
I check their related share library, and output as below:
arm-wrs-linux-gnueabi-readelf -d create_tpm_key

Dynamic section at offset 0x100c contains 25 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libcrypto.so.1.0.0]
 0x00000001 (NEEDED)                     Shared library: [libc.so.6]
 0x0000000c (INIT)                       0x10600
 0x0000000d (FINI)                       0x10af0
 0x00000019 (INIT_ARRAY)                 0x21000
 0x0000001b (INIT_ARRAYSZ)               4 (bytes)
 0x0000001a (FINI_ARRAY)                 0x21004
 0x0000001c (FINI_ARRAYSZ)               4 (bytes)
 0x6ffffef5 (GNU_HASH)                   0x10194
 0x00000005 (STRTAB)                     0x103a8
 0x00000006 (SYMTAB)                     0x10258
 0x0000000a (STRSZ)                      349 (bytes)
 0x0000000b (SYMENT)                     16 (bytes)
 0x00000015 (DEBUG)                      0x0
 0x00000003 (PLTGOT)                     0x210fc
 0x00000002 (PLTRELSZ)                   120 (bytes)
 0x00000014 (PLTREL)                     REL
 0x00000017 (JMPREL)                     0x10588
 0x00000011 (REL)                        0x10570
 0x00000012 (RELSZ)                      24 (bytes)
 0x00000013 (RELENT)                     8 (bytes)
 0x6ffffffe (VERNEED)                    0x10530
 0x6fffffff (VERNEEDNUM)                 2
 0x6ffffff0 (VERSYM)                     0x10506
 0x00000000 (NULL)                       0x0

arm-wrs-linux-gnueabi-readelf -d libtpm.so.0 

Dynamic section at offset 0x43dc contains 28 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libc.so.6]
 0x00000001 (NEEDED)                     Shared library: [libtspi.so.1]
 0x00000001 (NEEDED)                     Shared library: [libpthread.so.0]
 0x00000001 (NEEDED)                     Shared library: [libcrypto.so.1.0.0]
 0x0000000e (SONAME)                     Library soname: [libtpm.so.0]
 0x0000000c (INIT)                       0x1744
 0x0000000d (FINI)                       0x3eb0
 0x00000019 (INIT_ARRAY)                 0x14390
 0x0000001b (INIT_ARRAYSZ)               4 (bytes)
 0x0000001a (FINI_ARRAY)                 0x14394
 0x0000001c (FINI_ARRAYSZ)               4 (bytes)
 0x6ffffef5 (GNU_HASH)                   0xf8
 0x00000005 (STRTAB)                     0x90c
 0x00000006 (SYMTAB)                     0x1dc
 0x0000000a (STRSZ)                      2002 (bytes)
 0x0000000b (SYMENT)                     16 (bytes)
 0x00000003 (PLTGOT)                     0x144dc
 0x00000002 (PLTRELSZ)                   752 (bytes)
 0x00000014 (PLTREL)                     REL
 0x00000017 (JMPREL)                     0x1454
 0x00000011 (REL)                        0x1204
 0x00000012 (RELSZ)                      592 (bytes)
 0x00000013 (RELENT)                     8 (bytes)
 0x6ffffffe (VERNEED)                    0x11c4
 0x6fffffff (VERNEEDNUM)                 2
 0x6ffffff0 (VERSYM)                     0x10de
 0x6ffffffa (RELCOUNT)                   66
 0x00000000 (NULL)                       0x0

thanks,
Limeng

**Comment by jiazhang0** [comment]
So libcrypto is required in RDEPEND. 

I remeber libtspi.so and libtpm.so are provided by trousers right? If so, add it to RDEPEND as well.

**Comment by limeng-linux** [comment]
Hi Jia,

I modified code according to your review comment.

Thanks,
Limeng

File: recipes-tpm/openssl-tpm-engine/files/README [hunk 6]
```diff
@@ -0,0 +1,28 @@
+openssl-tpm-engine package provides tpm engine lib for openssl applications.
+
+==================================
+1. Examples of create_tpm_key tool 
+==================================
+(1). If tpm is not owned, run "tpm_takeownership -y -z".
+     This also set tpm SRK and Owner password to well-known key.
+     You can then run "tpm_changeownerauth -z -s -o" to set new SRK and Owner password.
```
Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/files/README]: s/"tpm_changeownerauth -z -s -o"/tpm_changeownerauth/

File: recipes-tpm/packagegroups/README [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+================================
+Create a self-signed cert using the TPM engine:
+
+1. Generate a TPM key and write it to a file:
+ $ create_tpm_key <keyfilename>
+
+2. Make the openssl certificate request:
+ $ openssl req -keyform engine -engine tpm -key <keyfilename> -new -x509 -days 365 -out <certfilename>
+
+Note: Detail description about openssl-tpm-engine, please refer to the README in source code.
```
Comment (jiazhang0) [inline recipes-tpm/packagegroups/README]: You need to move this content to feature/tpm/README.

File: recipes-tpm/openssl-tpm-engine/files/README [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+2. How to use tpm engine lib 
+==================================
+The name of tpm engine library is libtpm.so.
+There is an enhencement to the original opensource code.
+Add an additional way to pass SRK passwork to libtpm.so,
+that is using environment variable "TPM_SRK_PW".
+For example:
+	env TPM_SRK_PW=xxx openssl s_server ...
+
+NOTE: "env TPM_SRK_PW=#WELLKNOWN#" is used to pass well-known key.
```
Comment (jiazhang0) [inline recipes-tpm/openssl-tpm-engine/files/README]: You need to move this content to feature/tpm/README.

**Comment by limeng-linux** [comment]
Hi Jia,

modify over, please help to review when you are free.

thanks,
Limeng

**Comment by jiazhang0** [comment]
Tech approved.

**Pull request merged**